### DISCUSSION TITLE
How does proxy work with publish?

### DISCUSSION THREAD

**Discussion by EvanCarroll** [verdaccio/verdaccio#2669] [Q&A] [answered]
It's not clear from the documentation what a Verdaccio proxy is.

If I set up a proxy for `@acme/foo` and someone targets verdaccio trying to publish `@acme/foo` how do I configure the behavior and what is the default?

There are three obvious possibilities but none of them are covered in the docs,

1. The publish is reject
2. The publish is proxied.
3. The publish is done locally, and further requests are not proxied.

I would prefer the third option but when I try to publish to a package that is proxied I get 

```text
npm ERR! code EPUBLISHCONFLICT
npm ERR! publish fail Cannot publish over existing version.
npm ERR! publish fail Update the 'version' field in package.json and try again.
npm ERR! publish fail 
npm ERR! publish fail To automatically increment version numbers, see:
npm ERR! publish fail     npm help version

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/ecarroll/.npm/_logs/2021-11-11T18_45_39_813Z-debug.log
```

in npm

**Comment by edclement** [comment, marked as answer]
> There are three obvious possibilities but none of them are covered in the docs

I'll wait for someone else more familiar with Verdaccio's internal working to answer that. However, that error message indicates to me you are attempting to publish a package using the same version number as was previously published at some point. I'm assuming you're just running
```bash
npm publish
```

Has the version of the package (listed in `package.json`) you're trying to publish already been published to Verdaccio? According to the [npm publish](https://docs.npmjs.com/cli/v7/commands/npm-publish) docs:
> The publish will fail if the package name and version combination already exists in the specified registry.

> Once a package is published with a given name and version, that specific name and version combination can never be used again, even if it is removed with npm unpublish

That sounds like the issue you're running into.... I'm _assuming_ Verdaccio has implemented those same publishing rules.

**Reply by EvanCarroll** [reply]
No, but it's on the proxy which is I assume why Verdaccio is refusing to accept it. If I change the name in the `package.json` to `@acme/foobar` it will not be on the proxy and the publish will succeed and the package will exist on the verdaccio.

But again none of this is mentioned in the docs as far as I can see.

**Reply by juanpicado** [reply]
I'll complement what @edclement already explained. When you are trying to publish a `package@version` and you have `uplinks` defined (by default is `npmjs`) it will:
1. Ask to the defined uplink (in case you have the `proxy: npmjs` in your package access`) and download the manifest if exist, if does not will just don't persist anything (jump to step 3)
2. Will verify if in that manifest (in case step 1 is positive) if the version already exist.
3. If exist, will return 409 http status (http conflict) <-- what npmjs will display is `EPUBLISHCONFLICT` if not, will publish the package.

Here more info: https://verdaccio.org/docs/best#override-public-packages

In a nutshell, if the package access section has `proxy` it will always double check with the uplink and won't allow you have duplications.

**Reply by EvanCarroll** [reply]
That's a great explanation thanks.
